pkg/boot/fit/vfit.go

```diff
@@ -0,0 +1,211 @@
+// Copyright 2021 the u-root Authors. All rights reserved
+// Use of this source code is governed by a BSD-style
+// license that can be found in the LICENSE file.
+//
+// Performs signature checks on FDT images.
+// Currently supports PGP and raw PKCS1v15 RSA signatures.
+//
+// Expected FDT Format:
+//  Node: images
+//  Node: image_name
+//   P: data
+//   Node: signature*
+//    P: value
+//    P: algo          (ex. 'sha256,rsa4096', 'pgp')
+//    P: signer-name   (Optional)
+//    P: key-name-hint (Optional)
+//
+// vfit aims to be TOCTTOU-safe by reading files into memory before verifying.
+package fit
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/rsa"
+	"fmt"
+	"strings"
+	"unicode"
+
+	"github.com/u-root/u-root/pkg/dt"
+	"github.com/u-root/u-root/pkg/vfile"
+	"golang.org/x/crypto/openpgp"
+)
+
+var (
+	algs = map[string]crypto.Hash{
+		"MD4":       crypto.MD4,
+		"MD5":       crypto.MD5,
+		"SHA1":      crypto.SHA1,
+		"SHA224":    crypto.SHA224,
+		"SHA256":    crypto.SHA256,
+		"SHA384":    crypto.SHA384,
+		"SHA512":    crypto.SHA512,
+		"RIPEMD160": crypto.RIPEMD160,
+		"SHA3_224":  crypto.SHA3_224,
+		"SHA3_256":  crypto.SHA3_256,
+		"SHA3_384":  crypto.SHA3_384,
+		"SHA3_512":  crypto.SHA3_512,
+	}
+)
+
+type Signature interface {
+	fmt.Stringer
+	// Warning: If the signature does not exist or does not match the keyring,
+	// both the file and a signature error will be returned.
+	Verify([]byte, openpgp.KeyRing) (*bytes.Reader, error)
+}
+
+type PGPSignature struct {
+	name  string // Name of signature Node
+	value []byte
+	// Optional description fields
+	signer string
+	hint   string
+}
+
+type RSASignature struct {
+	name  string // Name of signature Node
+	hash  crypto.Hash
+	value []byte
+	// Optional description fields
+	signer string
+	hint   string
+}
+
+func (s PGPSignature) String() string {
+	return fmt.Sprintf("PGP Signature - name: %s, signer: '%s', hint: '%s'", s.name, s.signer, s.hint)
+}
+
+// Warning: If the signature does not exist or does not match the keyring,
+// both the file and a signature error will be returned.
+func (s PGPSignature) Verify(b []byte, ring openpgp.KeyRing) (*bytes.Reader, error) {
+	r := bytes.NewReader(b)
```

Comment:
If your objective is to be TOCTTOU-proof (see above conversation), then I think you may want to copy the data being passed in. My read of [bytes.Reader](https://cs.opensource.google/go/go/+/refs/tags/go1.17.1:src/bytes/reader.go) is that the `bytes.Reader` that you get back here is going to read from the same underlying array as `b`. Which seems not TOCTTOU proof. However, the threat model of the TOCTTOU is not really clear to me, the attacker is able to manipulate the contents of the array backing `b` but not the array backing whatever the return value gives back? (This is really more of a question for @hugelgupf and doesn't need to block this review.)

Reply:
I have no problem changing this, but since it is the current standard, I'll elect to leave it no-copy for the moment and allow @hugelgupf to later comment.
My hypothesis is this is a trade-off to save memory on tight systems.
See https://github.com/u-root/u-root/pull/2112 for current example work where we have to skip duplicating a file due to memory constraints on a ramfs.